third_party/move/move-vm/runtime/src/storage/struct_name_index_map.rs

```diff
@@ -1,43 +1,96 @@
 // Copyright (c) The Move Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+use claims::assert_none;
+use move_core_types::language_storage::{StructTag, TypeTag};
 use move_vm_types::loaded_data::runtime_types::{StructIdentifier, StructNameIndex};
 use parking_lot::RwLock;
-use std::collections::BTreeMap;
+use std::{collections::BTreeMap, sync::Arc};
 
 #[derive(Clone)]
 struct IndexMap<T: Clone + Ord> {
     forward_map: BTreeMap<T, usize>,
-    backward_map: Vec<T>,
+    backward_map: Vec<Arc<T>>,
 }
 
-/// A data structure to cache struct identifiers (address, module name, struct name) and
-/// use indices instead, to save on the memory consumption and avoid unnecessary cloning.
+/// A data structure to cache struct identifiers (address, module name, struct name) and use
+/// indices instead, to save on the memory consumption and avoid unnecessary cloning. It
+/// guarantees that the same struct name identifier always corresponds to a unique index.
 pub(crate) struct StructNameIndexMap(RwLock<IndexMap<StructIdentifier>>);
 
 impl StructNameIndexMap {
+    /// Returns an empty map with no entries.
     pub(crate) fn empty() -> Self {
         Self(RwLock::new(IndexMap {
             forward_map: BTreeMap::new(),
             backward_map: vec![],
         }))
     }
 
+    /// Maps the struct identifier into an index. If the identifier already exists returns the
+    /// corresponding index. This function guarantees that for any struct identifiers A and B,
+    /// if A == B, they have the same indices.
     pub(crate) fn struct_name_to_idx(&self, struct_name: StructIdentifier) -> StructNameIndex {
+        // Note that we take a write lock here once, instead of (*): taking a read lock, checking
+        // if the index is cached, re-acquiring the (write) lock, and checking again, as it makes
+        // things faster.
+        // Note that even if we do (*), we MUST check if another thread has cached the index before
+        // we reached this point for correctness. If we do not do this, we can end up evicting the
+        // previous index, and end up with multiple indices corresponding to the same struct. As
+        // indices are stored inside types, type comparison breaks!
         let mut index_map = self.0.write();
+
+        // Index is cached, return early.
         if let Some(idx) = index_map.forward_map.get(&struct_name) {
             return StructNameIndex(*idx);
         }
+
+        // Otherwise, the cache is locked and the struct name is not present. We simply add it
+        // to the cache and return the corresponding index.
         let idx = index_map.backward_map.len();
-        index_map.forward_map.insert(struct_name.clone(), idx);
-        index_map.backward_map.push(struct_name);
+        let prev_idx = index_map.forward_map.insert(struct_name.clone(), idx);
+        assert_none!(prev_idx, "Indexing map should never evict cached entries");
```

Comment:
we would panic here anyway, so as a matter of pattern, it might be better to unlock (drop the lock) first, then assert the invariant?

Reply:
done